sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-            lines.append((0, '{ rank = max; "'+root_node+'"}'))
-            
-            for obj in list(observed_obj): # for each variant of objective, add a link to the root node, and determine if it's sink
-                lines.append((0,'"'+translate(obj)+'" -> "'+root_node+'"'))
-                
-                sinkflag = False
-                for sink in sev_sinks:    
-                    if obj.split("|")[-1] == sink:
-                        sinkflag = True
-                        break
-                if sinkflag:
-                    lines.append((0,'"'+translate(obj)+'" [style="filled,dotted", fillcolor= salmon]'))
+            print('\t!!!! Objective', objective)
+            ag_name = objective.replace('|', '').replace('_', '').replace('-', '').replace('(', '').replace(')', '')
+            lines = ['digraph ' + ag_name + ' {',
+                     'rankdir="BT"; \n graph [ nodesep="0.1", ranksep="0.02"] \n node [ fontname=Arial, ' +
+                     'fontsize=24, penwidth=3]; \n edge [ fontname=Arial, fontsize=20,penwidth=5 ];']
+            root_node = _translate(objective, root=victim)
+            lines.append('"' + root_node + '" [shape=doubleoctagon, style=filled, fillcolor=salmon];')
+            lines.append('{ rank = max; "' + root_node + '"}')
+
+            # For each variant of objective, add a link to the root node, and determine if it's sink
+            for obj_variant in list(observed_obj):
+                lines.append('"' + _translate(obj_variant) + '" -> "' + root_node + '"')
+                if _is_severe_sink(obj_variant, sev_sinks):
+                    lines.append('"' + _translate(obj_variant) + '" [style="filled,dotted", fillcolor= salmon]')
                 else:
-                    lines.append((0,'"'+translate(obj)+'" [style=filled, fillcolor= salmon]'))
+                    lines.append('"' + _translate(obj_variant) + '" [style=filled, fillcolor= salmon]')
             
-            samerank = '{ rank=same; "'+ '" "'.join([translate(x) for x in observed_obj]) # all obj variants have the same rank
-            samerank += '"}'
-            lines.append((0,samerank))
-
+            # All obj variants have the same rank
+            lines.append('{ rank=same; "' + '" "'.join([_translate(obj) for obj in observed_obj]) + '"}')
 
+            node_signatures = dict()
             already_addressed = set()
-            for attackerID,attempts in team_level.items(): # for every attacker that obtains this objective
-                color = tcols[attackerID.split('-')[0]] # team color
-                ones = [''.join([action[0] for action in attempt]) for attempt in attempts]
-                unique = len(set(ones)) # count exactly unique attempts
-                #print(unique)
-                #print('team', attackerID, 'total paths', len(attempts), 'unique paths', unique, 'longest path:', max([len(x) for x in attempts]), \
```

Comment:
I think this part was used in the main paper, so I have kept it, but I have extracted it in a separate method, so that it is more readable and only one line has to be commented out